certbot-dns-bookmyname/certbot_dns_bookmyname/dns_bookmyname.py

```diff
@@ -0,0 +1,330 @@
+
+    def set_domain_zone(self, domain_id, zone):
+        """
+        Set the zone for a given domain name
+
+        NOTES: Extra whitespace within lines has no effect, but OTOH,
+        the zone cannot be entirely empty, so if it is, we replace it
+        with a single \n.
+        Also, TXT records (and possibly others) require double quotes.
+
+        This method will log in if necessary
+        """
+        if zone == '':
+            zone = '\n'
+        self.log_in()
+        try:
+            self.session.post(
+                "https://www.bookmyname.com/manager.cgi?cmd=gdp&mode=1",
+                {
+                    'did': domain_id,
+                    'gdp_zonefile': zone,
+                    'mode': 1,
+                    'Submit': 'Valider'
+                }
+            )
+        except:
+            raise ZoneError
+
+class _BookMyNameClient(object):
+    """
+    Encapsulates all communication with BookMyName.
+    """
+
+    def __init__(self, login, password, session):
+        self.manager = Manager(login, password, session)
+        self.domains_and_ids = {}
+
+    def login(self):
+        """
+        Log in and pre-fetch registered domains and ids
+
+        This method can be called several times; only the first one will
+        cause an actual login and domain+id list fetch.
+        """
+        if len(self.domains_and_ids) == 0:
+            try:
+                self.manager.log_in()
+                try:
+                    self.domains_and_ids = self.manager.get_domains_and_ids()
+                except Error as e:
+                    raise errors.PluginError(
+                        'Error getting BookMyName domains: {0}'.format(e)
+                    )
+            except Error as e:
+                raise errors.PluginError(
+                    'Error logging into BookMyName: {0}'.format(e)
+                )
+
+    def get_registered_domain(self, domain):
+        """
+        Return the registered domain matching the provided domain if any
+
+        A registered domain matches the provided domain if the provided
+        domain ends with the registered domain.
+
+        This method will log in if needed.
+        """
+        self.login()
+        for registered_domain in self.domains_and_ids:
+            if domain.endswith(registered_domain):
+                return registered_domain
+        raise errors.PluginError(
+            'No registered BookMyName domain match for domain {0}'
+            .format(domain)
+        )
+
+    def get_registered_domain_id(self, domain):
+        """
+        Return the ID of the matching registered domain if any
+
+        The ID is needed to pass to the website when setting the zone
+        contents (whereas getting the zone contents takes the domain,
+        not ID. Go figure).
+
+        This method will log in if needed.
+        """
+        self.login()
+        registered_domain = self.get_registered_domain(domain)
```

Comment:
Currently, it seems like we invoke `get_registered_domain` twice (once to get the domain, once here to get the ID — and we get both). This seems a little inefficient.

Could we have a single method that just returns the domain and its ID?

Reply:
Trying a few things, I'll get back to you as soon as I'm done.

Reply:
OK, I manually checked that in places where I use the id, I can also use the domain, so the id becomes unnecessary.

Reply:
Domain IDs are not used any more, only domain names.